### PR TITLE
Switch to use Patch() when pausing deployment configs

### PR DESCRIPTION
**Pull request by mfojtik**
Downstream part of upstream: https://github.com/kubernetes/kubernetes/pull/31220

### CODE REVIEW

**Comment by mfojtik** [comment]
[test]

@kargakis PTAL

File: pkg/cmd/util/clientcmd/factory.go [hunk 1]
```diff
@@ -510,12 +510,12 @@ func NewFactory(clientConfig kclientcmd.ClientConfig) *Factory {
 			if t.Spec.Paused {
 				return true, nil
 			}
-			t.Spec.Paused = true
+			body := []byte("{\"spec\":{\"paused\":true}}")
```
Comment (deads2k) [inline pkg/cmd/util/clientcmd/factory.go]: I think there's a function somewhere that generates the patch for you.

Comment (mfojtik) [inline pkg/cmd/util/clientcmd/factory.go]: somewhere in kube? I stoled this from the upstream PR https://github.com/kubernetes/kubernetes/pull/30919

Comment (mfojtik) [inline pkg/cmd/util/clientcmd/factory.go]: @deads2k I think CalculatePatches is an overkill for update simple like this, wdyt?

Comment (liggitt) [inline pkg/cmd/util/clientcmd/factory.go]: calculate patches means if the JSON serialization ever changes, we don't have to sweep our codebase for little embedded JSON patches

Comment (deads2k) [inline pkg/cmd/util/clientcmd/factory.go]: > @deads2k I think CalculatePatches is an overkill for update simple like this, wdyt?

I could see creating a more focused helper.  If you want this to become "standard", then it needs to be easy and maintainable.  I think that bugs caused by typos in string serialized json strategic merge patches will be nearly impossible to stamp out.

Comment (mfojtik) [inline pkg/cmd/util/clientcmd/factory.go]: @deads2k @liggitt agree, I can check if I can build something more minimal... I don't see an obvious way how to call set.CalculatePatches in here.

**Comment by kargakis** [comment]
You also need to update an extended test to use `oc rollout pause` now that we are switching to Patch:
https://github.com/openshift/origin/blob/b81f65c6fc829e387a130e5041ceba2b2a12e71a/test/extended/deployments/deployments.go#L454

**Comment by kargakis** [comment]
`oc rollout resume` actually

**Comment by mfojtik** [comment]
@kargakis PTAL, this refactor the upstream pause/resume to just pass the mutate function.

**Comment by mfojtik** [comment]
@kargakis tested and works, I will open upstream PR if you agree with the approach here.

@deads2k @liggitt fyi - no more random json strings

File: vendor/k8s.io/kubernetes/pkg/kubectl/cmd/rollout/rollout_pause.go [hunk 2]
```diff
@@ -119,17 +132,36 @@ func (o *PauseConfig) CompletePause(f *cmdutil.Factory, cmd *cobra.Command, out
 
 func (o PauseConfig) RunPause() error {
```
Comment (mfojtik) [inline vendor/k8s.io/kubernetes/pkg/kubectl/cmd/rollout/rollout_pause.go]: @deads2k @kargakis this and the resume are identical except the success message (paused/resumed) and the mutateFn. You think it is OK to a little copy here or a helper would be better?

File: pkg/cmd/util/clientcmd/factory.go [hunk 3]
```diff
@@ -503,40 +503,22 @@ func NewFactory(clientConfig kclientcmd.ClientConfig) *Factory {
...
 	}
 	w.PrintObjectSpecificMessage = func(obj runtime.Object, out io.Writer) {}
-	kPauseObjectFunc := w.Factory.PauseObject
-	w.Factory.PauseObject = func(object runtime.Object) (bool, error) {
-		switch t := object.(type) {
+	kPauseObjectFunc := w.Factory.PauseFn
+	w.Factory.PauseFn = func(object *resource.Info) (bool, error) {
+		switch t := object.Object.(type) {
 		case *deployapi.DeploymentConfig:
-			if t.Spec.Paused {
```
Comment (mfojtik) [inline pkg/cmd/util/clientcmd/factory.go]: @kargakis this is now one-liner

File: vendor/k8s.io/kubernetes/pkg/kubectl/cmd/rollout/rollout_pause.go (hunk 2 above)
Comment (mfojtik) [inline vendor/k8s.io/kubernetes/pkg/kubectl/cmd/rollout/rollout_pause.go]: I think copying is better here... I will have to make some generic option struct to make this work and I'm not sure it is worth (other than sharing this with `set`).

**Comment by kargakis** [comment]
Both Travis and Jenkins failures seem real

**Comment by ncdc** [comment]
Is this still for 1.4?

**Comment by mfojtik** [comment]
let's do this for 1.6 when we get Patch() in the clientset.

**Pull request closed without merging**